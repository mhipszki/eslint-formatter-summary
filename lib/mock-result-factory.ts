import { ESLint, Linter } from 'eslint';

type Rule = [string, Linter.LintMessage['severity']];

export default function resultFactory(failingRules: Rule[]): ESLint.LintResult {
  const messages = failingRules.map(function (rule) {
    const ruleId = rule[0];
    const severity = rule[1];
    return {
      ruleId,
      severity,
      message: `'${ruleId}' has failed`,
      line: 1,
      column: 1,
      nodeType: 'Identifier',
      source: "const foo = 'bar';",
    };
  });

  return {
    filePath: '',
    messages,
    suppressedMessages: [],
    errorCount: 1,
    fatalErrorCount: 0,
    warningCount: 0,
    fixableErrorCount: 0,
    fixableWarningCount: 0,
    usedDeprecatedRules: [],
  };
}
