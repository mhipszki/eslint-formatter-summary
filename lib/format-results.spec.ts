import { test, expect } from 'vitest';
import stripAnsi from 'strip-ansi';
import format from './format-results';
import { mockLintResult } from './mock-result-factory';

test('provides header, summary of rules and total', () => {
  const results = [
    mockLintResult([
      ['rule1', 1],
      ['rule2', 1],
    ]),
    mockLintResult([
      ['rule2', 1],
      ['rule3', 2],
    ]),
    mockLintResult([
      ['rule3', 2],
      ['rule4', 2],
      ['rule5', 2],
    ]),
  ];
  const output = stripAnsi(format(results, {}));
  expect(output).toMatchSnapshot();
});

test('defaults to sorting summary by rule id / ascending', () => {
  const results = [
    mockLintResult([
      ['B', 1],
      ['C', 2],
      ['A', 2],
    ]),
    mockLintResult([
      ['A', 1],
      ['C', 1],
    ]),
  ];
  expect(stripAnsi(format(results, {}))).toMatchSnapshot();
});

test('can sort summary by rule id', () => {
  const results = [
    mockLintResult([
      ['rule2', 1],
      ['rule3', 2],
    ]),
    mockLintResult([
      ['rule1', 1],
      ['rule2', 1],
    ]),
  ];
  expect(stripAnsi(format(results, { SORT_BY: 'rule' }))).toMatchSnapshot();
  expect(
    stripAnsi(format(results, { SORT_BY: 'rule', DESC: 'true' })),
  ).toMatchSnapshot();
});

test('can sort summary by num of errors', () => {
  const results = [
    mockLintResult([
      ['rule2', 2],
      ['rule3', 2],
    ]),
    mockLintResult([
      ['rule1', 1],
      ['rule2', 2],
    ]),
  ];
  expect(stripAnsi(format(results, { SORT_BY: 'errors' }))).toMatchSnapshot();
  expect(
    stripAnsi(format(results, { SORT_BY: 'errors', DESC: 'true' })),
  ).toMatchSnapshot();
});

test('can sort summary by num of warnings', () => {
  const results = [
    mockLintResult([
      ['rule2', 1],
      ['rule3', 2],
    ]),
    mockLintResult([
      ['rule1', 1],
      ['rule2', 1],
    ]),
  ];
  expect(stripAnsi(format(results, { SORT_BY: 'warnings' }))).toMatchSnapshot();
  expect(
    stripAnsi(format(results, { SORT_BY: 'warnings', DESC: 'true' })),
  ).toMatchSnapshot();
});

// test('omits result for ignored files', () => {
//   const ignoredFileResults: ESLint.LintResult = {
//     filePath: 'test.js',
//     messages: [
//       {
//         ruleId: undefined,
//         column: 1,
//         line: 1,
//         severity: 1,
//         message:
//           'File ignored because of a matching ignore pattern. Use "--no-ignore" to override.',
//       },
//     ],
//     suppressedMessages: [],
//     errorCount: 0,
//     fatalErrorCount: 0,
//     warningCount: 1,
//     fixableErrorCount: 0,
//     fixableWarningCount: 0,
//     usedDeprecatedRules: [],
//   };

//   expect(
//     stripAnsi(
//       format(
//         [
//           mockLintResult([
//             ['rule1', 1],
//             ['rule2', 2],
//           ]),
//           ignoredFileResults,
//         ],
//         {},
//       ),
//     ),
//   ).toMatchSnapshot();

//   expect(stripAnsi(format([ignoredFileResults], {}))).toMatchSnapshot();
// });
