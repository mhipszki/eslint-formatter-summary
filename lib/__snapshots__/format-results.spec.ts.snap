// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`formatResults > can sort summary by num of errors 1`] = `
" âœ¨  Summary of failing ESLint rules 
errors 0 warnings 1 rule: rule1
errors 1 warnings 0 rule: rule3
errors 2 warnings 0 rule: rule2
ðŸ”¥  4 problems in total (3 errors, 1 warnings)"
`;

exports[`formatResults > can sort summary by num of errors 2`] = `
" âœ¨  Summary of failing ESLint rules 
errors 2 warnings 0 rule: rule2
errors 1 warnings 0 rule: rule3
errors 0 warnings 1 rule: rule1
ðŸ”¥  4 problems in total (3 errors, 1 warnings)"
`;

exports[`formatResults > can sort summary by num of warnings 1`] = `
" âœ¨  Summary of failing ESLint rules 
errors 1 warnings 0 rule: rule3
errors 0 warnings 1 rule: rule1
errors 0 warnings 2 rule: rule2
ðŸ”¥  4 problems in total (1 errors, 3 warnings)"
`;

exports[`formatResults > can sort summary by num of warnings 2`] = `
" âœ¨  Summary of failing ESLint rules 
errors 0 warnings 2 rule: rule2
errors 0 warnings 1 rule: rule1
errors 1 warnings 0 rule: rule3
ðŸ”¥  4 problems in total (1 errors, 3 warnings)"
`;

exports[`formatResults > can sort summary by rule id 1`] = `
" âœ¨  Summary of failing ESLint rules 
errors 0 warnings 1 rule: rule1
errors 0 warnings 2 rule: rule2
errors 1 warnings 0 rule: rule3
ðŸ”¥  4 problems in total (1 errors, 3 warnings)"
`;

exports[`formatResults > can sort summary by rule id 2`] = `
" âœ¨  Summary of failing ESLint rules 
errors 1 warnings 0 rule: rule3
errors 0 warnings 2 rule: rule2
errors 0 warnings 1 rule: rule1
ðŸ”¥  4 problems in total (1 errors, 3 warnings)"
`;

exports[`formatResults > defaults to sorting summary by rule id / ascending 1`] = `
" âœ¨  Summary of failing ESLint rules 
errors 1 warnings 1 rule: A
errors 0 warnings 1 rule: B
errors 1 warnings 1 rule: C
ðŸ”¥  5 problems in total (2 errors, 3 warnings)"
`;

exports[`formatResults > provides header, summary of rules and total 1`] = `
" âœ¨  Summary of failing ESLint rules 
errors 0 warnings 1 rule: rule1
errors 0 warnings 2 rule: rule2
errors 2 warnings 0 rule: rule3
errors 1 warnings 0 rule: rule4
errors 1 warnings 0 rule: rule5
ðŸ”¥  7 problems in total (4 errors, 3 warnings)"
`;
